Add a config panel

pitchum opened the pull request:
The main goal of this PR is to address issue #38 

- Add config panel allowing tweaking some common prosody settings (like max upload size)
- Replace deprecated mod_http_upload with mod_http_file_share (needed to overcome some limitations)


## PR Status

- [x] Code finished and ready to be reviewed/tested
- [x] The fix/enhancement were manually tested by myself
- [ ] The fix/enhancement were manually tested by someone else

More detailed check-list:

- [x] `config_panel.toml` must be bullet-proof tested
  - a tiny typo can prevent the whole panel to show
  - all the parameters must already exist in prosody config, otherwise the config panel won't display
- [x] prevent conflicts between settings: `http_file_share_size_limit` should not be bigger than `http_file_share_global_quota`
- [x] auto-reload prosody after any change
- [ ] at some point we'll have to consider cleaning up the now obsolete directory `/var/xmpp-upload/`
- [ ] 🥵 big challenge: find a solution for people with multiple virtual hosts
- [x] known issue: custom settings will be overwritten on next app upgrade


## How to test

⚠️ Important: before considering testing this branch, please note that upgrading to this version will perform some irreversible changes. So make sure you have functional backups before upgrading.

1. Config files in `/etc/prosody/conf.d/` will be overwritten, so any custom parameters you may have set will be lost
2. shared files will be moved **and** renamed from `/var/xmpp-upload/..` to `/var/lib/prosody/...`

```
# fresh install
yunohost app install --force --debug \
 https://github.com/YunoHost-Apps/prosody_ynh/tree/add-config-panel

# or upgrading
yunohost app upgrade --force --debug prosody \
 --url https://github.com/YunoHost-Apps/prosody_ynh/tree/add-config-panel
```

pitchum committed:
Add config panel allowing tweaking some common prosody settings

pitchum committed:
Auto-reload prosody after changing settings in config panel

pitchum committed:
Replace deprecated mod_http_upload with mod_http_file_share

pitchum committed:
Bugfix: nginx does not have enough permissions to directly access files from mod_http_file_share.

anubister commented on scripts/_common.sh:
```diff
@@ -18,6 +18,9 @@ _configure_prosody() {
     ln -srf /etc/prosody/conf.avail/00.cfg.lua /etc/prosody/conf.d/
 
+    # Make sure data is migrated to the new http_file_share datastorage format
+    prosodyctl mod_migrate_http_upload xmpp-upload.${domain} ${domain}
```
Should be run only if Prosody was already installed. Crash in case of new installation.

anubister committed:
Update config_panel.toml

pitchum commented:
!gogogadgetoci

ericgaspar commented:
!testme

pitchum committed:
Rewrite available http_share option values...

- harmonize number decomposition format for clarity
- harmonize choices progressivity as much as possible (1/2/5...10/20/50...)
- include 2 infinity-like choices: 999TiB / 999years

pitchum committed:
Ensure custom settings persist after upgrading

pitchum commented:
!gogogadgetoci

anubister commented on config_panel.toml:
```diff
@@ -0,0 +1,68 @@
+        bind = ":/etc/prosody/conf.avail/__DOMAIN__.cfg.lua"
+
+        [main.upload.http_file_share_global_quota]
+        ask.en = "Maximum storage size per user"
```
storage for *all* users

pitchum committed:
fixup! Rewrite available http_share option values...
